### ISSUE TITLE
fido2-assert/fido2-cred do not support hmac-secret FIDO2 extension as opposed to examples

### ISSUE THREAD

**Issue by prusnak** [Yubico/libfido2#58]
It would be great if `fido2-assert`/`fido2-cred` supported hmac-secret FIDO2 extension the same way `examples/assert` and `example/cred` do, so one could integrate these standard tools into their workflows instead of building their own solution.

**Comment by martelletto** [comment]
tentative code pushed to [fido2-hmac](https://github.com/Yubico/libfido2/tree/fido2_hmac); apologies for the delay!

**Comment by prusnak** [comment]
Can you please provide `fido2-cred` and `fido2-assert` command lines so I can test?

Maybe these can become an example in manpages for fido2-cred/assert. Currently we have only usage without `-h` there.

**Comment by prusnak** [comment]
Can you please provide examples of how to use `fido2-cred`/`fido2-assert` for the hmac-secret extension scenario? Ideally this would be added to the `fido2-cred`/`fido2-assert` manpages as the second example.

**Comment by martelletto** [comment]
Sure. The steps are almost the same as the documented ones, with the difference that -h and a HMAC salt need to be present:

- Making a credential:

echo credential challenge | openssl sha256 -binary | base64 > cred_param
echo relying party >> cred_param
echo user name >> cred_param
dd if=/dev/urandom bs=1 count=32 | base64 >> cred_param
fido2-cred -M -hi cred_param /dev/hidraw7 | fido2-cred -V -ho cred

- Getting an assertion:

echo assertion challenge | openssl sha256 -binary | base64 > assert_param
echo relying party >> assert_param
head -1 cred >> assert_param
tail -n +2 cred > pubkey
dd if=/dev/urandom bs=1 count=64 | base64 -w0 >> assert_param <- hmac salt
fido2-assert -G -hi assert_param /dev/hidraw7 > /tmp/assert
fido2-assert -V -hi /tmp/assert pubkey es256
tail -1 /tmp/assert | base64 -d | xxd <- hmac secret

(Please note I have just amended the branch to add a -h option to fido2-assert -V, which was missing.)

**Comment by prusnak** [comment]
I tested the commands above (commit 0d7d1511d7e8ce97ebd58cff97344e3045afd430) with Trezor Model T and it fails on this line

```
fido2-assert -G -hi assert_param /dev/hidraw0 > /tmp/assert
```

with

```
fido2-assert: output error
```

**Comment by martelletto** [comment]
Thanks, I will take a look.

**Comment by martelletto** [comment]
Reverting #60 (9d44af2) fixes it for me. What are the device's expectation of the COSEAlgorithmIdentifier in this case?

**Comment by prusnak** [comment]
@andrewkozlik can you have a look please?

**Comment by andrewkozlik** [comment]
It works fine with the trezor-core master branch. The reason is that https://github.com/trezor/trezor-firmware/commit/9537bc40a58fb4e7c428fb7b7137e1fc2d004e6b never made it into the release, because the change was made too late after the code freeze. This commit is the equivalent of 9d44af2 in libfido2.

Note that the next day after further research I decided to disable the check completely to avoid compatibility issues, even though it's against the spec: https://github.com/trezor/trezor-firmware/commit/18998ff42f53cc022eefb619320cbdd19a956770

**Comment by prusnak** [comment]
Ah, thank you for the explanation Andrew!

I confirm that trezor-firmware master works with 0d7d1511d7e8ce97ebd58cff97344e3045afd430 (fido2-hmac branch HEAD) correctly!

We can close this issue after the fido2-hmac branch is merged to master. Thank you @martelletto for getting this done, much appreciated!

**Comment by martelletto** [comment]
Implemented in 183b556; thanks!
